solace/src/main/java/org/apache/apex/malhar/solace/AbstractSolaceGuaranteedIdempotentInputOperator.java

```diff
@@ -0,0 +1,649 @@
+  @Override
+  protected T processMessage(BytesXMLMessage message)
+  {
+
+    T payload = super.processMessage(message);
+    if (payload != null) {
+      currentWindowRecoveryState.put(message.getMessageIdLong(), payload);
+    }
+    recentMessage = message;
+    return payload;
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    //LOG.debug("Largest Recovery Wndow is : {} for current window: {}", idempotentStorageManager.getLargestRecoveryWindow(), windowId);
+    currentWindowId = windowId;
+    if (windowId <= idempotentStorageManager.getLargestRecoveryWindow()) {
+      //LOG.debug("About to handle recovery, current windowID is: {} largested recovered ID is: {}" + currentWindowId, idempotentStorageManager.getLargestRecoveryWindow());
+      handleRecovery(windowId);
+    } else {
+      if (super.consumer == null) {
+        super.startConsumer();
+        LOG.debug("Started Flow Consumer after recovery is complete");
+      }
+    }
+  }
+
+
+  @SuppressWarnings("unchecked")
+  protected void handleRecovery(long windowId)
+  {
+    LOG.info("++++++++++++++++++ Handle Recovery called");
+
+    Map<Long, T> recoveredData;
+    try {
+      recoveredData = (Map<Long, T>)idempotentStorageManager.load(operatorId, windowId);
+
+      if (recoveredData == null) {
+        return;
+      }
+      for (Map.Entry<Long, T> recoveredEntry : recoveredData.entrySet()) {
+        emitTuple(recoveredEntry.getValue());
+        inFlightRecoveryMessages.add(recoveredEntry.getValue());
+      }
+
+    } catch (IOException e) {
+      DTThrowable.rethrow(e);
+    }
+
+  }
+
+
+  @Override
+  public void endWindow()
+  {
+    @SuppressWarnings("unused")
+    boolean stateSaved = false;
+    boolean ackCompleted = false;
+    int messagesToAck = unackedMessages.size();
+
+    if (currentWindowId > idempotentStorageManager.getLargestRecoveryWindow()) {
+
+
+      if (recentMessage != null) {
+        lastMessages.put(currentWindowId, recentMessage);
+      }
+
+      try {
+
+        if (recentMessage != null) {
+          idempotentStorageManager.save(currentWindowRecoveryState, operatorId, currentWindowId);
+          stateSaved = true;
+          LOG.debug("Saved for window: " + currentWindowId);
+          currentWindowRecoveryState.clear();
+          LOG.debug("acking messages");
+          ackCompleted = ackMessages();
+          LOG.debug("Acked status: " + ackCompleted + " on window " + currentWindowId + " ack count: : " + messagesToAck);
+        }
+      } catch (Throwable t) {
+        if (!ackCompleted) {
+          LOG.info("confirm recovery of {} for {} does not exist", operatorId, currentWindowId, t);
+        }
+        DTThrowable.rethrow(t);
+      }
+
+
+      emitCount = 0; //reset emit count
+    } else {
+      currentWindowRecoveryState.clear();
+      ackCompleted = ackMessages();
+      LOG.debug("acking messages completed successfully: " + ackCompleted);
+    }
+
+  }
+
+
+  @Override
+  public void emitTuples()
```

Comment:
Can we break this 160 line function into smaller functions? My suggestions follow